src/Debugger/Impl/DataInspection/IRValueInfo.cs

```diff
@@ -0,0 +1,102 @@
+
+namespace Microsoft.R.DataInspection {
+    /// <summary>
+    /// Describes the result of evaluating an expression that produced a value that is not a promise or an active binding. 
+    /// </summary>
+    /// <remarks>
+    /// Note that most properties of the object will only have a meaningful value if the corresponding <see cref="RValueProperties"/>
+    /// flag was specified when producing the result. All properties which were not so requested will be <see langword="null"/>.
+    /// </remarks>
+    public interface IRValueInfo : IREvaluationInfo {
+        /// <summary>
+        /// String representation of the value.
+        /// </summary>
+        string Representation { get; }
+
+        /// <summary>
+        /// The kind of accessor that was used to obtain this <see cref="IRValueInfo"/> from its parent.
+        /// </summary>
+        RChildAccessorKind AccessorKind { get; }
+
+        /// <summary>
+        /// Type of the value, as computed by <c>typeof(...)</c>.
+        /// </summary>
+        string TypeName { get; }
+
+        /// <summary>
+        /// List of classes of the value, as computed by <c>classes(...)</c>.
+        /// </summary>
+        /// <remarks>
+        /// If <see cref="RValueProperties.Classes"/> was not specified, this property will be
+        /// <see langword="null"/>, rather than an empty collection.
+        /// </remarks>
+        IReadOnlyList<string> Classes { get; }
+
+        /// <summary>
+        /// Length of the value, as computed by <c>length(...)</c>.
+        /// </summary>
+        int? Length { get; }
+
+        /// <summary>
+        /// Number of attributes that this value has, as computed by <c>length(attributes(...))</c>.
+        /// </summary>
+        int? AttributeCount { get; }
+
+        /// <summary>
+        /// Number of slots that this value has, as computed by <c>length(slotNames(class(...)))</c>.
+        /// </summary>
+        int? SlotCount { get; }
+
+        /// <summary>
+        /// Number of names that the children of value have, as computed by <c>length(names(...))</c>.
+        /// </summary>
+        int? NameCount { get; }
+
+        /// <summary>
+        /// Dimensions that this value has, as computed by <c>dim(...)</c>.
+        /// </summary>
+        /// <remarks>
+        /// If <see cref="RValueProperties.Dim"/> was not specified, this property will be
+        /// <see langword="null"/>, rather than an empty collection.
+        /// </remarks>
+        IReadOnlyList<int> Dim { get; }
+
+        /// <summary>
+        /// Various miscellaneous flags describing this value.
+        /// </summary>
+        RValueFlags Flags { get; }
+
+        /// <summary>
+        /// <see langword="true"/> if <see cref="IREvaluationInfo.DescribeChildrenAsync"/> is known to return any items,
+        /// otherwise <see langword="false"/>.
+        /// </summary>
+        bool HasChildren { get; }
+    }
+
+    public static class RValueInfoExtensions {
```

Comment:
They are used inside, but are also intended to be used by API clients. My point is rather that they are really a part of the interface as far as clients should be concerned - the only reason why it's an extension class is because C# doesn't support default implementations for interface methods. 

---
In reply to: [62136331](https://github.com/Microsoft/RTVS/pull/1731#discussion_r62136331) [](ancestors = 62136331,62136058,62135595)